ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -134,15 +134,20 @@ public AmbryRequests(StoreManager storeManager, RequestResponseChannel requestRe
     this.connectionPool = connectionPool;
     this.metricRegistry = registry;
     this.serverConfig = serverConfig;
-    transformer = null;
-    if (serverConfig != null) {
-      try {
-        StoreKeyConverter keyConverter = storeKeyConverterFactory.getStoreKeyConverter();
-        transformer = Utils.getObj(serverConfig.serverMessageTransformer, storeKeyFactory, keyConverter);
-      } catch (Exception e) {
-        logger.error("Failed to create transformer", e);
+    /* All the request handlers share one single AmbryRequests object.
+     * But the StoreKeyConverter of the Transformer has a cache which shouldn't be shared among handlers.
+     * So use ThreadLocal transformer. */
+    this.transformer = ThreadLocal.withInitial(() -> {
```

Comment:
This works too, but I think the storeKeyConverter should be the thread local variable, not the transformer. Because the storekeyconverter is also used in other handlexxx methods and it's the one that has a local state that requires synchronization.

Reply:
In the AmbryRequests.java, we don't explicitly use keyConverter. 
We only provide the Transformer to the MessageSievingInputStream.transferInputStream. And MessageSievingInputStream.transferInputStream calls transformer.warmup only. Along the path, we don't touch keyConverter. 
Only the transformer LI implementation itself will use the keyConverter.


storeKeyconverter is not used in any other handleXXX.

Reply:
You are right.
Make keyConverter as well to avoid we create keyConverter every time when we handle one request.